torchtune/modules/common_utils.py

```diff
@@ -442,3 +448,138 @@ def delete_kv_caches(model: nn.Module):
             module.cache_enabled = False
             module.kv_cache = None
+
+
+def resize_token_embeddings(model: nn.Module, num_embeddings: int) -> None:
+    """
+    Resizes the token embeddings and the final output projection layer of a ``TransformerDecoder`` model.
+    The default init strategy is taking the mean of all the embeddings, new embeddings will be
+    instantiated to this value.
+
+    This function modifies the model in-place.
+
+    The primary purpose is to adjust the vocabulary size of a pre-trained model.
+    This is useful when fine-tuning a model on a dataset with a different
+    vocabulary or when adding special tokens.
+
+    Example:
+        >>> model = setup_model(...)
+        >>> tokenizer = setup_tokenizer(...)
+        >>> resize_token_embedding(model, tokenizer.num_embeddings)
+
+    Args:
+        model (nn.Module): The transformer model to modify. The model is
+            expected to have ``tok_embeddings`` (an ``nn.Embedding`` layer) and
+            ``output`` (e.g., ``nn.Linear`` or ``TiedLinear``) attributes.
+        num_embeddings (int): The desired number of embeddings in the resized
+            embedding layer and output projection layer.
+
+    Returns:
+        None: The function modifies the `model` in-place.
+
+    Raises:
+        AssertionError: When model not of type ``TransformerDecoder`` is passed in
+            and is a resize is needed. User should pass in `TransformerDecoder` part
+            of the model only.
+    """
+    need_resize = _need_to_resize(model, num_embeddings)
+    if not need_resize:
+        _log.info("Embedding resize not needed, skipping")
+        return
+
+    # Local import to break a circular dependency with TransformerDecoder module
+    from torchtune.modules.model_fusion import (
+        DeepFusionModel,
+        EarlyFusionModel,
+        FusionEmbedding,
+    )
+
+    if isinstance(model, DeepFusionModel) or isinstance(model, EarlyFusionModel):
+        model = model.decoder
+    if isinstance(model.tok_embeddings, FusionEmbedding) and need_resize:
+        raise AssertionError("Resize is not supported for FusionEmbedding")
+
+    is_fsdp = isinstance(model, FSDPModule)
+    mesh = getattr(model.tok_embeddings.weight, "device_mesh", None)
+    is_tied = isinstance(model.output, TiedLinear)
+
+    if is_fsdp:
+        model.tok_embeddings.unshard()
+        if not is_tied:
+            model.output.unshard()
+
+    _resize_token_embeddings_helper(model, num_embeddings)
+
+    if is_fsdp:
+        fully_shard(model.tok_embeddings, mesh=mesh)
+        if not is_tied:
+            fully_shard(model.output, mesh=mesh)
+
+
+def _need_to_resize(model: nn.Module, num_embeddings: int) -> bool:
+    try:
+        current_vocab = (
+            model.tok_embeddings.num_embeddings
+            if isinstance(model, TransformerDecoder)
+            else model.decoder.tok_embeddings.num_embeddings
+        )
```

Comment:
```suggestion
                current_vocab = getattr(model, "decoder", model).tok_embeddings.num_embeddings
```

Just a proposal, perhaps for some it's more difficult to read.

Reply:
Thanks - this was an old commit changed this function entirely, should be clearer